### ISSUE TITLE
Hello, I would like to ask if the data in onData can be used for the Media Source Buffer in MSE?

### ISSUE THREAD

**Issue by huotong1212** [Vanilagy/mp4-muxer#20]
constructor(
    onData: (data: Uint8Array, position: number) => void,
    onDone?: () => void,
    options?: { chunked?: true, chunkSize?: number }
);

Just like MSE loads fragmented MP4 (fMP4), can these media data be loaded in segments?

**Comment by Vanilagy** [comment]
In my other library, `webm-muxer`, I support exporting for live streaming and there's a [demo](https://vanilagy.github.io/webm-muxer/demo-streaming/) for it too. You can look at the code for it; I create a `MediaSource` and append buffers to it as I'm muxing.

This library here does not yet have support for a live streaming option, meaning the files will not play until they have been fully finalized. For live streaming support, I'd need to add support for fMP4 muxing. Not sure what the demand there is!

**Comment by huotong1212** [comment]
that's pretty good, but i found a little problem. After decoding an MP4 file (mime: video/mp4; codecs="avc1.640028"; profiles="iso5,iso6,mp41") to videoFrame and then encoding it to webm, vp9 format to add to the media source buffer, it can achieve continuous online playback, but the clarity has declined. I suspect this is due to the encoding, because even if I increase the bitrate, the encoded video is still not as clear as the MP4 format, but the entire MP4 file exported using mp4-muxer is clear. Is there a way for the media source buffer to read the decoded MP4 file directly?

**Comment by Vanilagy** [comment]
If you already have the entire MP4 file, why are you reencoding it as WebM just to stream it? Can't you just play back the original MP4 file? Could you specify in more detail what exactly you're trying to do?

Regarding the quality decline, I doubt this has anything to do with my muxers. Sounds like some encoder configuration thing.

**Comment by huotong1212** [comment]
Because I need to use TensorFlow on the web to pedict video frames, and then encode the output video frames into a new video to play in the browser. Thanks for your help ,is there any possible to do this thing?

**Comment by Vanilagy** [comment]
I'm sorry, I still don't quite understand what you're trying to do. But definitive answer is: Output from `mp4-muxer` cannot be piped into a MediaSource because it can't output fMP4. But `webm-muxer` can do it just fine - so maybe you should just use that!

**Comment by Vanilagy** [comment]
How's it looking?

**Comment by collin** [comment]
I am interested in being able to mux to fMP4. Do you have any bandwidth to tackle something like that? I can probably get some kind of budget for a bounty if you're amenable to that kind of thing.

**Comment by Vanilagy** [comment]
Whether I implement a big new feature into my muxers typically depends on how I allocate my free time, and lately that's been going into other activities. With a budget, I'd definitely prioritize implementing fMP4 - it's something I know will be a great addition to this library and will have to be done sooner or later. So yes, I'm definitely amenable to that :)

May I ask what you wish to use fMP4 support for? I need to figure out the actual usecase to better estimate the effort I'll need to put into this.

**Comment by collin** [comment]
Yeah, my use case is I have videos that I'm rendering in a headless browser. The ultimate destination for these is distributed in HLS/DASH manifests.

I would like to be able to render the segments out of order/in parallel. So I'd use one instance of a headless browser to render from say 0s -> 2s, and another from 2s ->4s. Then I'd mux and upload those segments directly to some file server to be served as part of a manifest.

So I'd need to be able to configure the muxer to set the correct sequence numbers and timestamps of the segments to be as-if they were part of the larger encode.

**Comment by Vanilagy** [comment]
I see, that's a cool usecase.

This will definitely take a larger effort to add, since I'll have to familiarize myself with fMP4, implement it and make sure it plays well with DASH. I'll be working closely with you to make sure your usecase works.

I say a fixed budget of $250 is fair given the effort this will require. Is that okay for you?

**Comment by collin** [comment]
I think that can work. I'll check to confirm and should be able to have an answer next week.

It's a been a little while since I've had my nose in the spec. I did at one point land a patch in mux.js to allow specifying arbitrary sequence numbers and timestamps. But their muxer expects your input data to be a mpegts stream, which I won't have available out of webcodecs.

My recollection is it was both easier and more difficult to figure out than I would have thought.

And the [dash reference player](dash reference player) was very helpful for verifying files work correctly.

**Comment by Vanilagy** [comment]
Please reach out once you have confirmation. 

That DASH reference player looks awesome, thanks! I need to see how I implement the global timestamp offset, since right now, timestamps are required to start at 0 because of the nature of the MP4 format. (not the case for WebM). But, I suppose that this is different for the fMP4 format.

**Comment by collin** [comment]
Alright, we're able to do this for a $250 bounty.

My understanding of the fragmented muxing is it's ~the same as the non-fragmented muxing, except all of the box names have a `f` variant. 

- `moov` -> `moof`
- `trak` -> `traf`
- `mvhd` -> `mfhd`
- etc.

_( If this is all review, apologies, just want to be as helpful as possible. )_

From an `mp4dump` of a fragmented file...

```ts
{
    "name": "moof",
    "header_size": 8,
    "size": 508,
    "children":[
```

The `mfhd` box includes the sequence number, which is one of the things we need to control:
```ts
    {
      "name": "mfhd",
      "header_size": 12,
      "size": 16,
      "sequence number": 3
    },
    {
      "name": "traf",
      "header_size": 8,
      "size": 484,
      "children":[
      {
        "name": "tfhd",
        "header_size": 12,
        "size": 32,
        "flags": 131130,
        "track ID": 1,
        "sample description index": 1,
        "default sample duration": 512,
        "default sample size": 653,
        "default sample flags": 16842752
      },
```
And then the `tfdt` box has the `base media decode time` which also needs to be controllable to do out of order muxing.
```ts
      {
        "name": "tfdt",
        "header_size": 12,
        "size": 20,
        "version": 1,
        "base media decode time": 25600
      },
      {
        "name": "trun",
        "header_size": 12,
        "size": 424,
        "version": 1,
        "flags": 2565,
        "sample count": 50,
        "data offset": 516,
        "first sample flags": 33554432
      }]
    }]
  },
  {
    "name": "mdat",
    "header_size": 8,
    "size": 1430
  },
  ```
  
  Other than that, it's roughly the same. The init fragment has a mostly typical `ftyp`/`moov`.
 
  
  To generate fragmented files `ffmpeg` is controlled by the `-movflags` I've typically used:
  
  ```sh
  ffmpeg -i INPUT.mp4 -c copy -movflags +cmaf+separate_moof OUTPUT.mp4
  ```
  
  The separate moof flag instructs ffmpeg to keep each stream in it's own moof/mdat pair. Not strictly required, but since I tend to store the streams separately anyway, I've habituated towards using that flag.

**Comment by Vanilagy** [comment]
Awesome, and thanks for the explanation - looking at and comparing box structures makes up basically 80% of development time of this library :P

Pretty sure I'll have time to start work on it this weekend.

**Comment by Vanilagy** [comment]
@collin How pressing is this feature to you right now? I'm still quite busy elsewhere.

**Comment by collin** [comment]
Hey, sorry I’ve had my head deep in other parts of my system.

This is still of interest to me, if it’s not something you can find into your schedule in the next week or so in may tackle it myself and open a PR.

**Comment by Vanilagy** [comment]
I think I'll have a good chance of fitting this in this weekend. I'll let you know once there's an update :) But good to know you're still interested - I've also been prioritizing other things.

**Comment by Vanilagy** [comment]
How well does your hacked-together semi-solution work? Are you able to parallelize encoding like you wanted, given that FFmpeg doesn't give you direct control of sequence numbers?

**Comment by Vanilagy** [comment]
@huotong1212 I just released v4.0.0 which adds support for fragmented MP4 files, which now also means we can use `mp4-muxer` together with MSE. I added a streaming demo which you can find at the top of the README.

@collin Definitely was way more effort than I expected to implement this properly. Required several refactors in the muxing code and a good amount of debugging to make sure the resulting file actually works properly with MSE. Fragments are automatically divided based on video key frames, so if you want 5-second video key frames, you simply encode one video key frame every 5 seconds. There's no support for manually setting sequence numbers / base media decode time yet, but I'd still be happy if you tried out the new fragmentation feature nonetheless. Regarding the sequence numbers, I need to reflect on how I'd prefer that to be done. For now, it's super niche (only you have requested it), so it might be worth monkey patching that in a fork instead. What I would still consider adding to the main library is different callbacks in the StreamTarget for fragments and init segments.

I'm closing this issue for now since the feature requested in the title now works. If there's more you need help with, don't mind writing here or opening a new issue :)
